datastore/elastic.go

```diff
@@ -0,0 +1,181 @@
+
+}
+
+func (e *ElasticDb) Post(shortBook models.ShortBook) (*models.Book, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return nil, initErr
+	}
+	ctx := context.Background()
+	_, updateErr := e.client.Update().Index(IndexName).Id(shortBook.Id).Doc(map[string]interface{}{"title": shortBook.Title}).Do(ctx)
+	if updateErr != nil {
+		return nil, updateErr
+	}
+	return e.Get(shortBook.Id)
+}
+
+func (e *ElasticDb) Delete(id string) error {
+	initErr := e.Init()
+	if initErr != nil {
+		return initErr
+	}
+	ctx := context.Background()
+	_, deleteErr := e.client.Delete().Index(IndexName).Id(id).Do(ctx)
+	return deleteErr
+}
+
+func (e *ElasticDb) queryBuilder(bookQuery models.BookQuery) (*elastic.BoolQuery, error) {
+	query := elastic.NewBoolQuery()
+	if bookQuery.Title != "" {
+		titleQuery := elastic.NewMatchQuery("title", bookQuery.Title)
+		query.Must(titleQuery)
+	}
+	if bookQuery.AuthorName != "" {
+		authorQuery := elastic.NewMatchQuery("author_name", bookQuery.AuthorName)
+		query.Must(authorQuery)
+	}
+	if bookQuery.PriceRange != "" {
+		rangeSlice := strings.Split(bookQuery.PriceRange, ",")
+		low, lowErr := strconv.ParseFloat(rangeSlice[0], 64)
+		high, highErr := strconv.ParseFloat(rangeSlice[1], 64)
+		if lowErr != nil || highErr != nil {
+			return nil, lowErr
+		}
+		rangeQuery := elastic.NewRangeQuery("price").Gte(low).Lte(high)
+		query.Must(rangeQuery)
+	}
+	return query, nil
+}
+
+func (e *ElasticDb) Search(bookQuery models.BookQuery) ([]models.Book, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return nil, initErr
+	}
+	ctx := context.Background()
+	var allBooks []models.Book
+	query, builderErr := e.queryBuilder(bookQuery)
+	if builderErr != nil {
+		return nil, builderErr
+	}
+	searchResult, searchErr := e.client.Search().Index(IndexName).Query(query).Size(QuerySize).Do(ctx)
+	if searchErr != nil {
+		return nil, searchErr
+	}
+	if searchResult.Hits.TotalHits.Value > 0 {
+		for _, hit := range searchResult.Hits.Hits {
+			var book models.Book
+			unmarshalErr := json.Unmarshal(hit.Source, &book)
+			if unmarshalErr != nil {
+				return nil, unmarshalErr
+			}
+			allBooks = append(allBooks, book)
+		}
+	}
+	return allBooks, nil
+
+}
+
+func (e *ElasticDb) GetStore() (models.Store, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return models.Store{}, initErr
+	}
+	ctx := context.Background()
+	distinctAuthorsAgg := elastic.NewCardinalityAggregation().Field("author_name.keyword")
+	allBooksSearchResults, searchErr := e.client.Search().Index(IndexName).Aggregation("bar", distinctAuthorsAgg).Size(0).Do(ctx)
+	if searchErr != nil {
+		return models.Store{}, searchErr
+	}
+	if allBooksSearchResults.Aggregations["bar"] == nil {
+		// Should handle this - use found
+	}
+
+	var numOfDistinctAuthors struct {
+		Count int64 `json:"value"`
+	}
+	unmarshalErr := json.Unmarshal(allBooksSearchResults.Aggregations["bar"], &numOfDistinctAuthors)
+	if unmarshalErr != nil {
+		return models.Store{}, unmarshalErr
+	}
```

Comment:
redundant - after checking if  `allBooksSearchResults.Aggregations["bar"]` exist, you can get the `count` valueת
Something like that:

```suggestion
     authorsCount, found := allBooksSearchResults.Aggregations["bar"]
     if !found {
     	return nil, &amp;elastic.Error{Status: http.StatusNotFound}
      }
       authors := authorsCount		 
       Books = allBooksSearchResults.Hits.TotalHits
```